src/main/java/gregtech/common/metatileentities/storage/MetaTileEntityQuantumStorageController.java

```diff
@@ -0,0 +1,439 @@
+        return false; // todo use mui2 for ui?
+    }
+
+    @Override
+    public boolean isValidFrontFacing(EnumFacing facing) {
+        return true;
+    }
+
+    @SuppressWarnings("SameParameterValue")
+    private IQuantumStorage<?> getStorage(BlockPos pos, boolean rebuild) {
+        if (getWorld().isRemote) return null;
+        if (storageInstances.containsKey(pos)) {
+            WeakReference<IQuantumStorage<?>> storageRef = storageInstances.get(pos);
+            IQuantumStorage<?> storage = storageRef.get();
+            if (storage != null) {
+                return storage;
+            }
+        }
+        // need to make sure it still exists
+        MetaTileEntity mte = GTUtility.getMetaTileEntity(getWorld(), pos);
+        if (mte instanceof IQuantumStorage<?>storage) {
+            storageInstances.put(pos, new WeakReference<>(storage));
+            return storage;
+        } else if (rebuild) {
+            // need to remove and rebuild
+            storagePositions.remove(pos);
+            rebuildNetwork();
+            return null;
+        }
+        return null;
+    }
+
+    @Override
+    public boolean canConnect(IQuantumStorage<?> storage) {
+        return !isDead && isInRange(storage.getPos());
+    }
+
+    private boolean isInRange(BlockPos pos) {
+        boolean isXValid = Math.abs(getPos().getX() - pos.getX()) <= MAX_DISTANCE_RADIUS;
+        boolean isYValid = Math.abs(getPos().getY() - pos.getY()) <= MAX_DISTANCE_RADIUS;
+        boolean isZValid = Math.abs(getPos().getZ() - pos.getZ()) <= MAX_DISTANCE_RADIUS;
+        return isXValid && isYValid && isZValid;
+    }
+
+    @Override
+    public void onRemoval() {
+        if (getWorld().isRemote) return;
+        isDead = true;
+        for (BlockPos pos : storagePositions) {
+            IQuantumStorage<?> storage = getStorage(pos, false);
+            if (storage != null) storage.setDisconnected();
+        }
+        handler.invalidate();
+    }
+
+    @Override
+    public void onPlacement(@Nullable EntityLivingBase placer) {
+        super.onPlacement(placer);
+        rebuildNetwork();
+    }
+
+    @Override
+    public void onLoad() {
+        calculateEnergyUsage();
+        super.onLoad();
+    }
+
+    // Used when this controller is initially placed. Try to find all possible
+    // storage instances that are connected and within our distance radius
+    // todo rework this to use neighbor cache somehow
+    @Override
+    public void rebuildNetwork() {
+        if (getWorld().isRemote) return;
+        var oldInstances = storageInstances;
+        var oldPositions = storagePositions;
+
+        storageInstances = new HashMap<>();
+        storagePositions = new HashSet<>();
+
+        Queue<BlockPos> searchQueue = new LinkedList<>();
+        Set<BlockPos> checked = new HashSet<>();
+
+        // check the posses around the controller
+        for (EnumFacing facing : EnumFacing.VALUES) {
+            searchQueue.add(getPos().offset(facing));
```

Comment:
To go off your todo comment about using the neighbor cache, this could first check `getNeighbor(facing)` is an instance of `IQuantumStorage`, and then only add the block pos to the searchQueue if that is true. However, I am unsure if the neighbor cache will be populated at this point, since this is called right after placement. That needs some checking.